Make parameter units more consistent

hamishwillee opened the pull request:
We check only some of our yaml parameter units for consistency - see https://github.com/PX4/Firmware/pull/15477#pullrequestreview-461351960 - 

Minor point, but we should probably check all of them. This is just a fixup to match the yaml values.

@bkueng I'd have fixed them all, and added validation, but the following aren't defined. You can merge and I'll stop thinking about this. 
Or you can confirm what you would like for some or all of these units we have listed I can fix those too:

{'S', '(m/s^2)^2', '(m/s)^2', 'SD', 'm/m', 's/(1000*PWM)', '(m/s^2)/s/sqrt(Hz)', 'Nm/(rad/s)', 'rad/sec', 'N', '(m/s)/s/sqrt(Hz)', '(m/sec**2)**2', 'amp-seconds per radian', '%m/s', 'Nm', 'Amps', 'henries', '1/s', 'N/(m/s)', '%alized_thrust', 'Index', 'Poles', 'm/s^2qrt(Hz)', 'rad*s*s/m', '(m/s)/(sqrt(hz))', 'Hz/s', 'kg*m*m', 'tan(rad)^2', 'us/C', 'rad/s**2', 'RPM/v', 'm/s / (rad/s)', 'n/a', 'rad*s/m', 'M', '%alized_thrust/s', 'A/%', 'm/s / (rad)', 'u', 'm^2', '%/rad/s', 'min'}

The specific params that the above are found in are:
```
Invalid unit: ASPD_BETA_GATE SD
Invalid unit: ctl_gain amp-seconds per radian
Invalid unit: ctl_start_rate Hz/s
Invalid unit: esc_index Index
Invalid unit: mot_i_max Amps
Invalid unit: mot_kv RPM/v
Invalid unit: mot_ls henries
Invalid unit: mot_num_poles Poles
Invalid unit: ASPD_BETA_GATE SD
Invalid unit: ASPD_SC_P_NOISE 1/s
Invalid unit: ASPD_TAS_GATE SD
Invalid unit: BAT_N_CELLS S
Invalid unit: COM_EF_C2T A/%
Invalid unit: EKF2_BARO_GATE SD
Invalid unit: EKF2_BETA_GATE SD
Invalid unit: EKF2_DRAG_NOISE (m/sec**2)**2
Invalid unit: EKF2_EVP_GATE SD
Invalid unit: EKF2_EVV_GATE SD
Invalid unit: EKF2_GBIAS_INIT rad/sec
Invalid unit: EKF2_GND_EFF_DZ M
Invalid unit: EKF2_GND_MAX_HGT M
Invalid unit: EKF2_GPS_P_GATE SD
Invalid unit: EKF2_GPS_V_GATE SD
Invalid unit: EKF2_GYR_B_NOISE rad/s**2
Invalid unit: EKF2_HDG_GATE SD
Invalid unit: EKF2_MAG_GATE SD
Invalid unit: EKF2_OF_GATE SD
Invalid unit: EKF2_RNG_A_IGATE SD
Invalid unit: EKF2_RNG_GATE SD
Invalid unit: EKF2_RNG_SFE m/m
Invalid unit: EKF2_TAS_GATE SD
Invalid unit: EKF2_TERR_GRAD m/m
Invalid unit: FW_PR_FF %/rad/s
Invalid unit: FW_PR_P %/rad/s
Invalid unit: FW_RR_FF %/rad/s
Invalid unit: FW_RR_P %/rad/s
Invalid unit: FW_WR_FF %/rad/s
Invalid unit: FW_WR_P %/rad/s
Invalid unit: FW_YR_FF %/rad/s
Invalid unit: FW_YR_P %/rad/s
Invalid unit: NAV_FT_FS n/a
Invalid unit: NAV_FT_RS n/a
Invalid unit: HTE_ACC_GATE SD
Invalid unit: HTE_HT_ERR_INIT %alized_thrust
Invalid unit: HTE_HT_NOISE %alized_thrust/s
Invalid unit: LTEST_ACC_UNC (m/s^2)^2
Invalid unit: LTEST_MEAS_UNC tan(rad)^2
Invalid unit: LTEST_POS_UNC_IN m^2
Invalid unit: LTEST_VEL_UNC_IN (m/s)^2
Invalid unit: LPE_FLW_R m/s / (rad)
Invalid unit: LPE_FLW_RR m/s / (rad/s)
Invalid unit: LPE_LT_COV m^2
Invalid unit: LPE_PN_B (m/s^2)/s/sqrt(Hz)
Invalid unit: LPE_PN_P m/s^2qrt(Hz)
Invalid unit: LPE_PN_T (m/s)/(sqrt(hz))
Invalid unit: LPE_PN_V (m/s)/s/sqrt(Hz)
Invalid unit: MC_PITCH_P 1/s
Invalid unit: MC_ROLL_P 1/s
Invalid unit: MC_YAW_P 1/s
Invalid unit: MC_YAW_WEIGHT 1/s
Invalid unit: MOT_SLEW_MAX s/(1000*PWM)
Invalid unit: GND_SPEED_D %m/s
Invalid unit: GND_SPEED_I %m/s
Invalid unit: GND_SPEED_IMAX %m/s
Invalid unit: GND_SPEED_P %m/s
Invalid unit: GND_SPEED_THR_SC %m/s
Invalid unit: SDLOG_UTC_OFFSET min
Invalid unit: INA226_CONFIG u
Invalid unit: SENS_IMU_TEMP_I us/C
Invalid unit: SENS_IMU_TEMP_P us/C
Invalid unit: SIH_IXX kg*m*m
Invalid unit: SIH_IXY kg*m*m
Invalid unit: SIH_IXZ kg*m*m
Invalid unit: SIH_IYY kg*m*m
Invalid unit: SIH_IYZ kg*m*m
Invalid unit: SIH_IZZ kg*m*m
Invalid unit: SIH_KDV N/(m/s)
Invalid unit: SIH_KDW Nm/(rad/s)
Invalid unit: SIH_Q_MAX Nm
Invalid unit: SIH_T_MAX N
Invalid unit: VT_B_DEC_FF rad*s*s/m
Invalid unit: VT_B_DEC_I rad*s/m
Invalid unit: WV_GAIN 1/s
```

bkueng commented:
Nice cleanup, thanks. I think we want those based on SI units as well.
`norm` -> `%`, but I didn't check all of them.

hamishwillee commented:
@bkueng I fixed up norm and updated the list in the description. I'm sure I can fix a few of these up to SI units - e.g Henries. But not sure what we'd do for SD (standard deviation), "Index", "Poles", "min", "(m/s)/(sqrt(hz))" etc. "S" as a unit for batteries might be technically wrong, but it is also right :-)

bkueng commented:
You're right, it does make sense. Just that we'll end up having many of them. Do they go into mavlink spec as well?

> But not sure what we'd do for SD (standard deviation), "Index", "Poles", "min", "(m/s)/(sqrt(hz))" etc. "S" as a unit for batteries might be technically wrong, but it is also right :-)

They should also be very clear from the param description.

Either way is fine with me, removing or adding to the standard set.

BazookaJoe1900 commented on src/modules/mc_hover_thrust_estimator/hover_thrust_estimator_params.c:
```diff
@@ -49,7 +49,7 @@
  * @min 0.0001
  * @max 1.0
- * @unit normalized_thrust/s
+ * @unit %alized_thrust/s
```
this is not that clear

hamishwillee replied:
That is a search replace error. Will Fix. Thanks @BazookaJoe1900 

@bkueng FYI, I am reverting back to `normalized_thrust/s`. Is there a better unit?

bkueng replied:
Hmm let's ask @bresch what he wants as he's the author of all of these params.

bresch replied:
We could use %/s but then we need to scale the result by 100.
If we can keep `normalized_thrust/s` I'm happy with it.

dagar commented on src/drivers/heater/heater_params.c:
```diff
@@ -53,7 +53,7 @@ PARAM_DEFINE_INT32(SENS_TEMP_ID, 0);
  * @category system
  * @group Sensors
- * @unit C
+ * @unit celcius
```
celsius?
```suggestion
 * @unit celsius
```

hamishwillee commented on src/modules/mc_hover_thrust_estimator/hover_thrust_estimator_params.c:
```diff
@@ -77,7 +77,7 @@ PARAM_DEFINE_FLOAT(HTE_ACC_GATE, 3.0);
  * @min 0.0
  * @max 1.0
- * @unit normalized_thrust
+ * @unit %alized_thrust
```
@bkueng I think this one can just be "%" right? The fact that it is normalised thrust is implied by the description.

bresch replied:
@hamishwillee Yes, we can use % but then we need to scale the result `0-100% -> 0.0-1.0`. Is that automatically done in QGC?

hamishwillee replied:
@bresch The simple answer is that I don't know. For now I have reverted this and we can revisit in a later round.

@bkueng This affects everything else we changed from "norm". I guess we should hold off merging until we know (unless you know?)

bkueng replied:
Scaling is done on the UI side, so no other change required.

bresch replied:
![Screen Capture_select-area_20200814100721](https://user-images.githubusercontent.com/14822839/90228133-fdaeeb00-de15-11ea-8a6f-f4c0a38875fc.png)
Hmm, it seems that the `@unit norm` is automatically converted and displayed as `%` in QGC

hamishwillee commented:
@bkueng 

Re https://github.com/PX4/Firmware/pull/15502#issuecomment-670356295 Here is my proposal on what we add to checker, what we change, what we delete, and some questions.  If you confirm/answer I'll make the changes. No huge urgency :-)

Add these ones the the checker:
- `rad/sec` (is SI unit)
- `(m/s^2)^2`
- `(m/s)^2`
- `tan(rad)^2`
- `m^2`
- `N`
- `Nm`
- `N/(m/s)`
- `us/C`
- `%/rad/s`
- `%m/s`
- `A/%`
- `Nm/(rad/s)`
- `1/s`
- `Hz/s`
- `min` (minutes)
- `S`. Note this is for battery as in 4S battery. Is this OK?
- `s/(1000PWM)` - (for [MOT_SLEW_MAX](https://dev.px4.io/master/en/advanced/parameter_reference.html#MOT_SLEW_MAX))


Suggest change these and add to checker (confirm)
- `normalized_thrust` to `%` (assume what is normalised is defined by the param docs)
- `radss/m` to `rad/m` (assume is typo)
-  `rads/m` to `rad/m`
- `amp-seconds per radian` to `C/rad` - (coulomb/rad) that OK?
- `m/s / (rad)` to `m/s/rad`
- `henries` to `H`
- `rad/s**2` to `rad/s^2`
- `Amps` to `A`
- `m/s / (rad/s)` to  `m/s/(rad/s)`
- `kg*m*m` to `kg m^2` - OK? - it is SIH_MASS unit of vehicle inertia.
- `RPM/v` to `rpm/V`

Propose delete units (not needed):
- `Index` is for [esc_index](https://dev.px4.io/master/en/advanced/parameter_reference.html#esc_index). Propose remove.
- `Poles is for [mot_num_poles](https://dev.px4.io/master/en/advanced/parameter_reference.html#mot_num_poles). Propose remove units
- `n/a`. This is implied.

Questions
- EKF2_GND_EFF_DZ and EKF2_GND_MAX_HGT have unit `M`. This is typo right? It is just `m`.
- unit for INA226_CONFIG is "u". What is that supposed to be?
- LPE_PN_P `m/s^2qrt(Hz)` - that has to be a typo. Would it be `(m/s^2) sqrt(Hz)`?
- What about these two (do we reduce):
  - `(m/s^2)/s/sqrt(Hz)` - should this be `(m/s^3/sqrt(Hz)`
  - `(m/s)/s/sqrt(Hz)` - should this be `(m/s^2)/sqrt(Hz)`
- There are a couple of gradient units (e.g. EKF2_TERR_GRAD) in m/m. Techically this is unitless, but I propose to add it anyway. OK?
- `SD` indicates standard deviation, which is useful to know, but you should know that from the description. Technically "The standard deviation is always represented by the same unit of measurement as the variable in question.". So I suggest I just change this for the specific cases- OK?

bkueng commented:
`1/s = Hz`
`Hz/s` -> `1/s^2` would be a bit more canonical.
`m/s/(rad/s)` -> simplify to `m/rad`?

> EKF2_GND_EFF_DZ and EKF2_GND_MAX_HGT have unit M. This is typo right? It is just m.

Right

> unit for INA226_CONFIG is "u". What is that supposed to be?

It's a register config, you can drop the unit. I don't even know why this is a param.

> LPE_PN_P m/s^2qrt(Hz) - that has to be a typo. Would it be (m/s^2) sqrt(Hz)?

This is probably an artifact from your search-and-replace. I see it as `m/s/sqrt(Hz)`.

> What about these two (do we reduce): 

I would.

The rest looks good to me, thanks Hamish.

hamishwillee commented:
Thanks @bkueng . I think all done except I decided to keep "SD" and I wasn't sure what to do with `normalized_thrust/s`?

I also updated the yaml with these values. If we decide to accept SD and add something for `normalized_thrust/s` I could further add validation code in the python to fail if anything else incorrect is added.

hamishwillee commented:
@bkueng FYI, also added validation check for param units. The set matches yaml. 

- I included SD - wasn't clear to me in all cases the original units.
- I included normalized_thrust, and normalized_thrust/s - until resolves with bresch.

So this is waiting now on Don to confirm whether QGC is smart enough to know that a max/min of 1 indicates that values should be translated to this from GUI. If not, then we would have to revert all the norm cases.

hamishwillee commented:
@bkueng @bresch I reverted the changes from % to `norm` (and added it to the "allowed" lists of units). 
- The `norm` clearly "works:.
- I can't find a % to test - the only one with values between 0 and 1 is DSHOT_MIN. @bkueng You might want to change that to norm to guarantee that it actually works properly in QGC.

I would rather get this in and change norm later than wait on fixes/changes to QGC handling.

@bkueng for review/merge when you get back from holiday

bkueng reviewed (approved):
Thanks @hamishwillee. If the % is not working, I'd rather fix it on the QGC side instead of changing to norm.
